File: python-algorithm/leetcode/problem_590.py
"""590. N-ary Tree Postorder Traversal
https://leetcode.com/problems/n-ary-tree-postorder-traversal/
"""
from typing import List


class Solution:
    def postorder(self, root: 'Node') -> List[int]:
        if not root:
            return []
        ans = []
        for node in root.children:
            ans.extend(self.postorder(node))
        ans.append(root.val)
        return ans

    def postorder2(self, root: 'Node') -> List[int]:
        if not root:
            return []
        ans = []
        q = [(0, root)]  # 0 means the count of visited children of root
        while q:
            cnt, node = q.pop()
            if not node:
                continue
            if cnt == len(node.children):
                ans.append(node.val)
            if cnt < len(node.children):
                q.append((cnt + 1, node))
                q.append((0, node.children[cnt]))
        return ans


class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children

File: java-algo/src/main/java/com/leetcode/Problem590.java
package com.leetcode;

import java.util.ArrayList;
import java.util.List;

/**
 * 590. N-ary Tree Postorder Traversal
 * https://leetcode.com/problems/n-ary-tree-postorder-traversal/
 */
public class Problem590 {
    public List<Integer> postorder(Node root) {
        List<Integer> ans = new ArrayList<>();
        if (root == null) {
            return ans;
        }
        for (Node child : root.children) {
            List<Integer> list = postorder(child);
            ans.addAll(list);
        }
        ans.add(root.val);
        return ans;
    }

    static class Node {
        public int val;
        public List<Node> children;

        public Node() {
        }

        public Node(int _val) {
            val = _val;
        }

        public Node(int _val, List<Node> _children) {
            val = _val;
            children = _children;
        }
    }

    ;
}
